### PR TITLE
Aligner/nemotron5

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

place to make all changes to support nemotron5 training
**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 1]
```diff
@@ -1078,53 +1083,14 @@ def dummy():
...
-                    # create tar file
-                    if self.pack_nemo_file:
-                        self._make_nemo_file_from_folder(save_path, tmpdir)
-                    else:
-                        # Get the folder path from the save_path and move all values inside the tmpdir to the folder
-                        folder_path = os.path.dirname(save_path)
-
-                        for file in os.listdir(tmpdir):
-                            shutil.move(os.path.join(tmpdir, file), folder_path)
+                tmpdir = dist_ckpt_dir.parent
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:1086]: why is the whole previous code removed?
I'm pretty sure we can't do that

File: nemo/utils/callbacks/nemo_model_checkpoint.py [hunk 2]
```diff
@@ -240,11 +252,18 @@ def on_save_checkpoint(self, trainer, pl_module, checkpoint):
...
                 torch.distributed.barrier()
             backup_path = self._backup_existing_nemo_ckpt(trainer)
-            pl_module.save_to(save_path=app_state.model_restore_path)
-            logging.info(f"New .nemo model saved to: {app_state.model_restore_path}")
-        if backup_path is not None and is_global_rank_zero():
-            logging.info(f'Removing old .nemo backup {backup_path}')
-            get_filesystem(backup_path).rm(backup_path)
+            if not self.remove_backup_path:
+                consumed_samples = trainer.fit_loop.epoch_loop.batch_progress.total.processed
+                pl_module.save_to(
```
Comment (mikolajblaz) [inline nemo/utils/callbacks/nemo_model_checkpoint.py:257]: Why the new model is saved or not based on `remove_backup_path` flag?

File: nemo/utils/callbacks/nemo_model_checkpoint.py [hunk 3]
```diff
@@ -93,6 +95,16 @@ def __init__(
             logging.debug("Checking previous runs")
             self.nemo_topk_check_previous_run()
 
+    def _save_topk_checkpoint(self, *args, **kwargs):
+        trainer = args[0]
+        prev_save_top_k = trainer.strategy.save_top_k
+
+        trainer.strategy.save_top_k = True
+        output = super()._save_topk_checkpoint(*args, **kwargs)
```
Comment (mikolajblaz) [inline nemo/utils/callbacks/nemo_model_checkpoint.py:103]: wrap in 
```
try:
    output = super()._save_topk_checkpoint(*args, **kwargs)
finally:
    trainer.strategy.save_top_k = prev_save_top_k
```

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 4]
```diff
@@ -387,29 +388,33 @@ def save_checkpoint(
...
-            # converts the optimizer states to their sharded equivalents
-            sharded_optim_state = self.optimizer_sharded_state_dict(
-                unsharded_optim_state=checkpoint['optimizer_states'][0]
-            )
-
-            # Check whether to save optim states
-            include_optimizer = True if not storage_options else storage_options.get('include_optimizer', True)
-            if include_optimizer:
-                checkpoint['optimizer_states'] = [sharded_optim_state]
+            if self.save_top_k:
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:391]: I would suggest moving this logic out of the strategy, instead do the differentiation regular/.nemo save in the `save_checkpoint` method of the NeMoModelCallback. Otherwise there is so much entanglement in the logic.

**Review by mikolajblaz — CHANGES_REQUESTED** [review]
Left some comments, but it would be good for @dimapihtar to look as well since it changes NeMo 1 checkpointing logic

**Review by dimapihtar — COMMENTED** [review]
can we run this PR once it's ready with nemo-ci nightly to make sure it doesn't break anything? Because I'm seeing a lot of checkpinting changes which can probably lead to ci failures.

File: nemo/utils/callbacks/nemo_model_checkpoint.py (hunk 3 above)
Comment (arendu) [reply nemo/utils/callbacks/nemo_model_checkpoint.py:103]: cc @gshennvm

File: nemo/utils/callbacks/nemo_model_checkpoint.py (hunk 2 above)
Comment (arendu) [reply nemo/utils/callbacks/nemo_model_checkpoint.py:257]: cc @gshennvm

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 1 above)
Comment (arendu) [reply nemo/collections/nlp/parts/nlp_overrides.py:1086]: We've been using this branch with no issues. I'll let @gshennvm  comment further.
